data_model/primitives/src/small.rs

```diff
@@ -0,0 +1,259 @@
+        /// string slice. This is not an implementation of [`FromStr`](core::str::FromStr),
+        /// because the latter implies **fallible** conversion, while this
+        /// particular conversion is **infallible**.
+        pub fn from_str(other: &str) -> Self {
+            Self(SmallString::from_str(other))
+        }
+
+        #[inline]
+        /// Checks if the specified pattern is the prefix of given string.
+        pub fn starts_with(&self, pattern: &str) -> bool {
+            self.0.starts_with(pattern)
+        }
+    }
+
+    impl<A: Array<Item = u8>> From<SmallString<A>> for SmallStr {
+        fn from(string: SmallString<A>) -> Self {
+            Self(SmallString::from_str(SmallString::as_str(&string)))
+        }
+    }
+
+    impl IntoSchema for SmallStr {
+        fn type_name() -> String {
+            String::type_name()
+        }
+        fn schema(map: &mut MetaMap) {
+            String::schema(map);
+        }
+    }
+}
+
+mod small_vector {
+    use super::*;
+
+    /// Wrapper struct around [`smallvec::SmallVec`] type. Keeps `N`
+    /// elements on the stack if `self.len()` is less than `N`, if not,
+    /// produces a heap-allocated vector.
+    ///
+    /// To instantiate a vector with `N` stack elements,
+    /// ```ignore
+    /// use iroha_data_model::small::SmallVec;
+    ///
+    /// let a: SmallVec<[u8; 24]> = SmallVec(smallvec::smallvec![32]);
+    /// ```
+    pub struct SmallVec<A: Array>(pub smallvec::SmallVec<A>);
+
+    impl<A: Array> Default for SmallVec<A> {
+        fn default() -> Self {
+            Self(smallvec::SmallVec::new())
+        }
+    }
+
+    impl<A: Array> Clone for SmallVec<A>
+    where
+        A::Item: Clone,
+    {
+        fn clone(&self) -> Self {
+            Self(self.0.clone())
+        }
+    }
+
+    impl<A: Array> fmt::Debug for SmallVec<A>
+    where
+        A::Item: fmt::Debug,
+    {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_tuple("SmallVec").field(&self.0).finish()
+        }
+    }
+
+    impl<A: Array> FromIterator<A::Item> for SmallVec<A> {
+        fn from_iter<T: IntoIterator<Item = A::Item>>(iter: T) -> Self {
+            Self(iter.into_iter().collect())
+        }
+    }
+
+    impl<A: Array> Serialize for SmallVec<A>
+    where
+        A::Item: Serialize,
+    {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            Serialize::serialize(&self.0, serializer)
+        }
+    }
+
+    impl<'de, A: Array> Deserialize<'de> for SmallVec<A>
+    where
+        A::Item: Deserialize<'de>,
+    {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            deserialize_from_smallvec(deserializer)
+        }
+    }
+
+    impl<A: Array> PartialEq for SmallVec<A>
+    where
+        A::Item: PartialEq + Eq,
```

Comment:
```suggestion
        A::Item: Eq,
```

should be implied?

Reply:
It should be `PartialEq`. Not `Eq` here.